For review

hex-for-hecate opened the pull request.

Commit:
Ignore package-lock and client secrets files

Commit:
Set up basic access to the Google Sheets API

Commit:
Reflect on the form of a "Nexus adaptation" and on the topology of Nexus systems more broadly

Commit:
Mock up some more realistic sheet grades

Commit:
Mock up a realistic demo

Commit:
With help of some pointers from @colinbdclark and @BlueSlug, reorganize and flesh out grades.

Commit:
Implement working demo of cell and clientSession with help from @cindyli

Commit:
Cleanup and add TODOs

Commit:
Move API scopes to external configuration

Commit:
Work on whole-spreadsheet component

Commit:
Successfully fetch all data from a spreadsheet

Commit:
Deprecate imperative test code

Commit:
Ignore VSCode debugger configuration

Commit:
Fix linter errors

Commit:
Update notes and TODOs

Commit:
Move and rename test and example code to clarify their purposes

Commit:
Add basic writing to remote sheets, move TODOs from source code to README, rename a bunch of things

amb26 commented on src/spreadsheets.js:
```diff
@@ -0,0 +1,286 @@
+spreadsheets.spreadsheet.getRemoteContent = function(that, apiClient) {
+    var metadataPromise = apiClient.getSpreadsheet(apiClient, that.options.spreadsheetId);
+    var requests = [];
+    metadataPromise.then(function(res) {
```
As we discussed, this could be broken up somewhat by being rexpressed as a transforming promise chain.

amb26 commented on src/spreadsheets.js:
```diff
@@ -0,0 +1,286 @@
+    modelListeners: {
+        '*': {
+            func: 'spreadsheets.spreadsheet.setRemoteContent',
+            args: ['{that}', '{apiClient}', '{change}.path.0', '{change}.value'],
```
Looks like you are aware that the "*" style produces will produce one message per sheet, with the first path segment allocated to this, but thought I would double-check. In future we might want more fine-grained binding. Note that there is a very inefficient utility fluid.model.diff that you could make use of quickly enough https://github.com/amb26/infusion/blob/FLUID-6145/src/framework/core/js/DataBinding.js#L1584 - or go for something more heavyweight and specialised in future.

amb26 commented on src/spreadsheets.js:
```diff
@@ -0,0 +1,286 @@
+    var togo;
+    for (i = str.length - 1; i >= 0; i--) {
+        var char = str[i];
+        if (char !== 'q') {
```
Some explanation of the strategy here would be helpful. I think you yourself would also appreciate some test cases for this algo and also split it off into its own file.

amb26 commented on src/spreadsheets.js:
```diff
@@ -0,0 +1,286 @@
+        'onCreate.getRemoteContent': 'spreadsheets.spreadsheet.getRemoteContent({that}, {apiClient})'
+    },
+    // When the model is written to, update the remote sheet
+    // TODO: where is the change source stored?
```
Change sources are stored in the framework's "transaction" object when a change is in progress - you can see it being accessed here - change.transaction.sources  - the transaction goes by as the 5th argument to a programmatic change listener assembled here - https://github.com/amb26/infusion/blob/FLUID-6145/src/framework/core/js/DataBinding.js#L1689 - but you will probably not need it beyond the "excludeSource" definition on setRemoteContent

amb26 commented on README.md:
```diff
@@ -1,2 +1,73 @@
+what is my ideal setup process?
+open the demo, possibly open a google spreadsheet
+
+It seems like we'd need to manually implement polling to get updates in response to sheet content changing.
```
Looks like in the long term we probably could get value from the drive push notification API - https://developers.google.com/drive/api/v3/reference/changes/watch - the restriction that all it could do is call back to a publically exposed HTTP endpoint in the long run won't be too onerous given we will have to host some kind of publically usable service in any case.

amb26 commented on src/spreadsheets.js:
```diff
@@ -0,0 +1,286 @@
+        valuePromise.then(function(res) {
+            fluid.each(res.data.valueRanges, function({range, values}) {
+                var sheetName = range.split('!')[0];
+                that.applier.change(sheetName, values);
```
May as well apply these in a transaction, even though it's perhaps unlikely that someone might want to depend on atomicity of updates who is listening to stuff in multiple sheets. https://github.com/amb26/infusion/blob/FLUID-6145/tests/framework-tests/core/js/DataBindingTests.js#L410

amb26 commented on src/examples.js:
```diff
@@ -0,0 +1,115 @@
+// I suppose I can make the POST request from the browser 
+// here's what I don't understand, is there an Infusion-y way to request
+// that a Nexus server creates a certain component?
+// How would Infusion know that a particular component should live in the Nexus
```
A good approach might be the "Visible Nexus" one where you simply set up a designated component root which represents that all of its descendents will be mirrored.

amb26 commented on src/examples.js:
```diff
@@ -0,0 +1,115 @@
+// and only exist here as a mirrored one?
+// I guess it could theoretically be done through a mixin grade, like
+// nexusClient.mirroredComponent
+// that might require a differentiation between local gradeNames and remote gradeNames
```
The "Visible Nexus" did this abominable hack - https://github.com/amb26/fluid-authoring/blob/FLUID-4884/src/client/js/ComponentGraphRemote.js#L33 also relying on the fact (probably a historical blunder) that any unknown grade would be treated as an empty grade. This ensured that the local grade would not actually attempt to instantiate - in the "new framework" perhaps a more civilised approach would be to direct the framework to only instantiate the tree to the point of component shells using the "breakAt" directive - https://github.com/amb26/infusion/blob/FLUID-6145/src/framework/core/js/FluidIoC.js#L2318

Commit:
Separate the A1 coordinate conversion algorithm, test it, and revise to pass tests

Commit:
Update URLs to reflect move to fluid-project

Commit:
Rewrite tests to use jqUnit

Commit:
Explain conversion algorithm

Commit:
Switch expected and actual values in test

Commit:
Add node-jqunit dev dependency

Commit:
Add state of development notes

Commit:
Refactor spreadsheets.spreadsheet

Moved the asynchronous control flow for the initial data fetching to a transforming promise chain.

Commit:
Add linting configuration
